### ISSUE TITLE
Flushing handlers in role's tasks/main.yml has no effect.

### ISSUE THREAD

**Issue by x-moe-x** [ansible/ansible#76998]
### Summary

**Problem:**
I have a role that relies on handlers being flushed before continuing with the play. This used to work in past versions of Ansible but fails to work on my current machine.

**Setup:**

Host System:
My system is Debian 11 64 bit, using Ansible 2.12.2.
(It used to work under Debian 10, using Ansible 2.9.27)
Guest: Vagrant based Debian 9 VM, provisioned with Ansible

**Question:**
This behavior seems to be wrong to me. What can I do to achieve the desired execution order?

### Issue Type

Bug Report

### Component Name

meta

### Ansible Version

```console
$ ansible --version
ansible [core 2.12.2]
  config file = /home/moe/WebstormProjects/he-server/ansible.cfg
  configured module search path = ['/home/moe/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/lib/python3/dist-packages/ansible
  ansible collection location = /home/moe/.ansible/collections:/usr/share/ansible/collections
  executable location = /usr/bin/ansible
  python version = 3.9.2 (default, Feb 28 2021, 17:03:44) [GCC 10.2.1 20210110]
  jinja version = 2.11.3
  libyaml = True
```

### Configuration

```console
$ ansible-config dump --only-changed
CALLBACKS_ENABLED(/home/moe/WebstormProjects/he-server/ansible.cfg) = ['profile_tasks', 'timer']
DEFAULT_FORCE_HANDLERS(/home/moe/WebstormProjects/he-server/ansible.cfg) = True
DEFAULT_HOST_LIST(/home/moe/WebstormProjects/he-server/ansible.cfg) = ['/home/moe/WebstormProjects/he-server/staging']
DEFAULT_LOG_PATH(/home/moe/WebstormProjects/he-server/ansible.cfg) = /home/moe/WebstormProjects/he-server/execution.log
DEFAULT_ROLES_PATH(/home/moe/WebstormProjects/he-server/ansible.cfg) = ['/home/moe/WebstormProjects/he-server/roles/build', '/home/moe/WebstormProjects/he-server/roles/misc', '/home/moe/WebstormProjects/he-server/roles/server', '/home/moe/WebstormProjects/he-server/roles/shared', '/home/moe/WebstormProjects/he-se>
```

### OS / Environment

Debian 11, Buster 64 bit

### Steps to Reproduce

```yaml (paste below)
---
- name: perform complicated thing
  block:
    - when: some_condition
      block:
        - name: some task
          ...

        - name: important task
          debug: msg="important"
          changed_when: true
          notify:
            - some handler 1
            - some handler 2

# this seems to be ignored
- meta: flush_handlers

# this is executed before the handlers are called
- name: other task that relies on handlers being executed
  ...

```

### Expected Results

Expected execution order:

1. some task
2. some handler 1
3. some handler 2
4. other task that relies on handlers being executed

### Actual Results

```console
Observed execution order:

1. some task
2. other task that relies on handlers being executed -> fails

After the task failing the handlers are invoked and the Play ends prematurely:

1. some handler 1
2. some handler 2
```

### Code of Conduct

- [X] I agree to follow the Ansible Code of Conduct

**Comment by mattclay** [comment]
@x-moe-x Can you provide the playbook output with `-vvv` to show the issue?

**Comment by x-moe-x** [comment]
https://pastebin.com/kMvNhBL9

**Comment by sivel** [comment]
With the reproducer you've provided, I cannot actually reproduce this.  `- meta: flush_handlers` works as expected from a task.

```
TASK [test : some task] ****************************************************************
ok: [localhost] => {
    "msg": "Hello world!"
}

TASK [test : important task] ***********************************************************
changed: [localhost] => {
    "msg": "important"
}

TASK [test : meta] *********************************************************************

RUNNING HANDLER [test : some handler 1] ************************************************
ok: [localhost] => {
    "msg": "handler 1"
}

RUNNING HANDLER [test : some handler 2] ************************************************
ok: [localhost] => {
    "msg": "handler 2"
}

TASK [test : other task that relies on handlers being executed] ************************
ok: [localhost] => {
    "msg": "Hello world!"
}
```

Please provide an complete small reproducer that you have run that causes this issue, that we can also run.  And attach the verbose output of this reproducer.

needs_info

**Comment by x-moe-x** [comment]
Ok, I think I understand what the problem is. I was able to reproduce a working minimal example on my machine
```---
- name: important task
  tags: [ specific_tag ]
  debug: msg="important task"
  changed_when: true
  notify:
    - some handler 1
    - some handler 2

- meta: flush_handlers

- name: other task that relies on handlers being executed
  tags: [ specific_tag ]
  debug: msg="other task that relies on handlers being executed"```

It obviously has to do with the tags I put in place. When I call the play with the `specific_tag` I get the handlers being flushed only at the end. However when I add this tag also to the meta task, the play runs with the handlers in correct order.

This is odd however as it used to work without the need for tagging the flush_task in the past, almost since day one of the script. I only ran into problems after updating from Debian 10 to 11, as well as updating Ansible from 2.9.27 to 2.12.2

**Comment by sivel** [comment]
Ok, this sounds expected then.  Your `meta` task doesn't have tags applied to it, so if you limit to `specific_tag`, the `meta` task will not run.

You may want to add the `always` tag to the `meta` task.

If you have further questions please stop by IRC or the mailing list:

   * IRC: #ansible on irc.libera.chat
   * mailing list: https://groups.google.com/forum/#!forum/ansible-project

**Comment by bcoca** [comment]
that was a bug that we fixed in #67508
